File: backend/app/api/audit_api.py
from typing import Optional, List, Dict, Any
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database.connection import get_db
from app.services.audit_service import audit_service
from pydantic import BaseModel, ConfigDict
from datetime import datetime

router = APIRouter(prefix="/audit", tags=["Module 9: System Audit Vault"])


class AuditLogSchema(BaseModel):
    id: int
    session_uuid: Optional[str] = None
    module_id: str
    action: str
    severity: str
    operator_role: str
    source_node: Optional[str] = None
    destination_node: Optional[str] = None
    details: Dict[str, Any]
    timestamp: datetime

    model_config = ConfigDict(from_attributes=True)


class CreateAuditLogRequest(BaseModel):
    module_id: str
    action: str
    severity: str = "INFO"
    session_uuid: Optional[str] = None
    operator_role: str = "GRID_ADMIN"
    source_node: Optional[str] = None
    destination_node: Optional[str] = None
    details: Optional[Dict[str, Any]] = None


@router.get("/logs", response_model=List[AuditLogSchema])
async def get_audit_logs(
    module_id: Optional[str] = Query(None, description="Filter by module ID or ALL"),
    severity: Optional[str] = Query(None, description="Filter by severity level (INFO, SUCCESS, WARNING, CRITICAL)"),
    session_uuid: Optional[str] = Query(None, description="Filter by QKD session UUID"),
    search: Optional[str] = Query(None, description="Free text search on action, nodes, role"),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db)
):
    """Retrieves paginated audit log entries from Supabase PostgreSQL."""
    logs = await audit_service.get_logs(
        module_id=module_id,
        severity=severity,
        session_uuid=session_uuid,
        search=search,
        limit=limit,
        offset=offset,
        db=db
    )
    return logs


@router.get("/stats")
async def get_audit_stats(db: AsyncSession = Depends(get_db)):
    """Computes module breakdown and severity summary statistics across all system audit logs."""
    return await audit_service.get_stats(db=db)


@router.post("/log", response_model=AuditLogSchema)
async def create_audit_log(
    req: CreateAuditLogRequest,
    db: AsyncSession = Depends(get_db)
):
    """Records a new system audit log entry into Supabase PostgreSQL."""
    log = await audit_service.log_event(
        module_id=req.module_id,
        action=req.action,
        severity=req.severity,
        session_uuid=req.session_uuid,
        operator_role=req.operator_role,
        source_node=req.source_node,
        destination_node=req.destination_node,
        details=req.details,
        db=db
    )
    return log

File: frontend/src/services/auditApi.ts
export interface SystemAuditLogItem {
  id: number;
  session_uuid?: string;
  module_id: string;
  action: string;
  severity: 'INFO' | 'SUCCESS' | 'WARNING' | 'CRITICAL';
  operator_role: string;
  source_node?: string;
  destination_node?: string;
  details: Record<string, any>;
  timestamp: string;
}

export interface SystemAuditStats {
  total_events: number;
  critical_events: number;
  warning_events: number;
  success_events: number;
  info_events: number;
  module_breakdown: Record<string, number>;
  success_rate_pct: number;
}

const API_BASE = '/api/v1/audit';

export async function fetchAuditLogs(params?: {
  module_id?: string;
  severity?: string;
  session_uuid?: string;
  search?: string;
  limit?: number;
  offset?: number;
}): Promise<SystemAuditLogItem[]> {
  const query = new URLSearchParams();
  if (params?.module_id) query.append('module_id', params.module_id);
  if (params?.severity) query.append('severity', params.severity);
  if (params?.session_uuid) query.append('session_uuid', params.session_uuid);
  if (params?.search) query.append('search', params.search);
  if (params?.limit) query.append('limit', params.limit.toString());
  if (params?.offset) query.append('offset', params.offset.toString());

  const response = await fetch(`${API_BASE}/logs?${query.toString()}`);
  if (!response.ok) {
    throw new Error('Failed to fetch system audit logs');
  }
  return response.json();
}

export async function fetchAuditStats(): Promise<SystemAuditStats> {
  const response = await fetch(`${API_BASE}/stats`);
  if (!response.ok) {
    throw new Error('Failed to fetch audit stats');
  }
  return response.json();
}
